File: java/code/src/com/suse/manager/model/products/Channel.java
/**
 * Copyright (c) 2013 SUSE LLC
 *
 * This software is licensed to you under the GNU General Public License,
 * version 2 (GPLv2). There is NO WARRANTY for this software, express or
 * implied, including the implied warranties of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
 * along with this software; if not, see
 * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
 *
 * Red Hat trademarks are not licensed under GPLv2. No permission is
 * granted to use or replicate Red Hat trademarks that are incorporated
 * in this software or its documentation.
 */

package com.suse.manager.model.products;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.apache.commons.lang3.builder.ToStringBuilder;
import org.simpleframework.xml.Attribute;

/**
 * A software Channel in a Product.
 */
public class Channel {

    /** Status attributed to channels that are installed. */
    public static final String STATUS_PROVIDED = "P";
    /** Status attributed to channels that are not installed. */
    public static final String STATUS_NOT_PROVIDED = ".";

    /** The label. */
    @Attribute
    private String label;

    /** The status. */
    @Attribute
    private String status;

    @Attribute
    private boolean installerUpdates;

    /**
     * Default constructor.
     */
    public Channel() {
        // required by Simple XML
    }

    /**
     * Instantiates a new channel.
     * @param labelIn the label in
     * @param statusIn the status in
     * @param installerUpdatesIn is an installer update channel
     */
    public Channel(String labelIn, String statusIn, boolean installerUpdatesIn) {
        label = labelIn;
        status = statusIn;
        installerUpdates = installerUpdatesIn;
    }

    /**
     * Gets the label.
     * @return the label
     */
    public String getLabel() {
        return label;
    }

    /**
     * Gets the status.
     * @return the status
     */
    public String getStatus() {
        return status;
    }

    /**
     * @return true if it is a installer update repo
     */
    public boolean isInstallerUpdates() {
        return installerUpdates;
    }

    /**
     * Check if this channel is provided according to mgr-ncc-sync (P).
     * @return true if this channel is provided, otherwise false
     */
    public boolean isProvided() {
        return STATUS_PROVIDED.equals(status);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean equals(Object other) {
        if (!(other instanceof Channel)) {
            return false;
        }
        Channel otherChannel = (Channel) other;
        return new EqualsBuilder()
            .append(getLabel(), otherChannel.getLabel())
            .append(getStatus(), otherChannel.getStatus())
            .isEquals();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int hashCode() {
        return new HashCodeBuilder()
            .append(getLabel())
            .append(getStatus())
            .toHashCode();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString() {
        return new ToStringBuilder(this)
        .append("label", getLabel())
        .append("status", getStatus())
        .toString();
    }

    /**
     * Set the product label
     * @param labelIn label
     */
    public void setLabel(String labelIn) {
        this.label = labelIn;
    }

    /**
     * Set the product status (eg. P or . )
     * @param statusIn status
     */
    public void setStatus(String statusIn) {
        this.status = statusIn;
    }
}

File: susemanager/src/mgr_sync/channel.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2014 SUSE
#
# This software is licensed to you under the GNU General Public License,
# version 2 (GPLv2). There is NO WARRANTY for this software, express or
# implied, including the implied warranties of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
# along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
#
# SUSE trademarks are not licensed under GPLv2. No permission is
# granted to use or replicate SUSE trademarks that are incorporated
# in this software or its documentation.

from enum import Enum

class Channel(object):  # pylint: disable=too-many-instance-attributes

    class Status(str, Enum):  # pylint: disable=too-few-public-methods
        INSTALLED = "INSTALLED"
        AVAILABLE = "AVAILABLE"
        UNAVAILABLE = "UNAVAILABLE"

    def __init__(self, data):
        self.base_channel = data['parent'] == 'BASE'
        self.summary = data["summary"]
        self.label = data["label"]
        self.name = data["name"]
        self.parent = data["parent"]
        self.status = data["status"]
        self.url = data["source_url"]
        self.optional = data["optional"]
        self.installer_updates = data["installer_updates"]
        self._children = []

    def __repr__(self):
        return self.to_ascii_row(True)

    @property
    def short_status(self):
        if self.status == Channel.Status.AVAILABLE:
            return "[ ]"
        else:
            return "[%s]" % str(self.status[0])

    def summary_or_url(self):
        return (self.summary + "").strip() or self.url or "N/A"

    def add_child(self, channel):
        self._children.append(channel)

    @property
    def children(self):
        return sorted(self._children, key=lambda child: child.label)

    def to_ascii_row(self, compact=False):
        if not compact:
            return "{0} {1} {2} [{3}]".format(
                self.short_status,
                self.name,
                self.summary_or_url(),
                self.label)
        else:
            return "{0} {1}".format(self.short_status, self.label)


def parse_channels(data, log):
    """
    Parses the data returned by SUSE Manager list channels API.
    Returns a dictionary where keys are the base channels labels and values
    are instances of Channel. Each instance is a base channel with all its
    "child" channels associated.
    """

    log.info("Parsing channels...")

    channels = {}

    for bc in [entry for entry in data if entry['parent'] == 'BASE']:  # pylint: disable=invalid-name
        base_channel = Channel(bc)
        log.debug("Found base channel '{0}'".format(base_channel.name))
        channels[base_channel.label] = base_channel

    for entry in data:
        if entry['parent'] == 'BASE':
            continue
        channel = Channel(entry)
        log.debug("Found channel '{0}'".format(channel.name))
        if channel.parent in channels:
            channels[channel.parent].add_child(channel)
        else:
            log.debug("Base Channel '{0}' undefined".format(channel.parent))

    return channels


def find_channel_by_label(label, channels, log):
    """ Looks for channel with label
    :param channels: a data structure returned by `parse_channels`
    :param label: the label to search
    :return: None if the channel is not found, a Channel instance otherwise
    """

    log.info("Searching for channels with label '{0}'".format(label))

    if label in list(channels.keys()):
        log.debug("Found '{0}'".format(channels[label]))
        return channels[label]

    for bc in list(channels.values()):  # pylint: disable=invalid-name
        matches = [c for c in bc.children if c.label == label]
        if len(matches) == 1:
            log.debug("Found '{0}'".format(matches[0]))
            return matches[0]

    return None
